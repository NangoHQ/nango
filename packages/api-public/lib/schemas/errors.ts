import * as z from 'zod';

import { apiSchemaRegistry } from './schema.js';

import type { FastifyReply, RawServerDefault, RouteGenericInterface } from 'fastify';
import type { ZodTypeProvider } from 'fastify-type-provider-zod';
import type { IncomingMessage, ServerResponse } from 'http';

export const schemaNotFound = z.object({
    error: z.strictObject({
        code: z.literal('not_found'),
        errorId: z.uuid().optional(),
        message: z.string()
    })
});

export const schemaServerError = z.object({
    error: z.strictObject({
        code: z.literal('server_error'),
        errorId: z.uuid().optional(),
        message: z.string()
    })
});

export const schemaUnauthorized = z.object({
    error: z.strictObject({
        code: z.literal('unauthorized'),
        message: z.string()
    })
});

apiSchemaRegistry.add(schemaNotFound, { id: 'ResponseNotFound' });
apiSchemaRegistry.add(schemaServerError, { id: 'ResponseServerError' });
apiSchemaRegistry.add(schemaUnauthorized, { id: 'ResponseUnauthorized' });

type FastifyReplyError<TCode extends number, TError extends z.ZodObject<any>> = FastifyReply<
    RouteGenericInterface,
    RawServerDefault,
    IncomingMessage,
    ServerResponse,
    unknown,
    {
        readonly response: Readonly<Record<TCode, TError>>;
    },
    ZodTypeProvider
>;

export async function resNotFound(res: FastifyReplyError<404, typeof schemaNotFound>, message?: string): Promise<void> {
    const err: z.infer<typeof schemaNotFound> = {
        error: {
            code: 'not_found',
            message: message ?? 'Not Found'
        }
    };
    return res.status(404).send(err);
}

export async function resServerError(res: FastifyReplyError<500, typeof schemaServerError>, message: string, errorId?: string): Promise<void> {
    const err: z.infer<typeof schemaServerError> = {
        error: {
            code: 'server_error',
            message,
            errorId
        }
    };
    return res.status(500).send(err);
}

export async function resUnauthorized(res: FastifyReplyError<401, typeof schemaUnauthorized>, message: string): Promise<void> {
    const err: z.infer<typeof schemaUnauthorized> = {
        error: {
            code: 'unauthorized',
            message: message ?? 'Unauthorized'
        }
    };
    return res.status(401).send(err);
}
