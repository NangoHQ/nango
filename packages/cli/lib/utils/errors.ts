import type { ErrorObject } from 'ajv';
import chalk from 'chalk';
import { inspect } from 'util';

export function displayValidationError({ data, validation, model }: { data: any; validation: (Error | ErrorObject)[]; model?: string }) {
    if (model) {
        console.log(chalk.blue('model:'), model);
    }
    console.log(chalk.blue('data:'), inspect(data, false, 5, true));
    console.log('');
    for (const error of validation) {
        if (error instanceof Error) {
            console.log('-', error.message);
            continue;
        }

        const paths = error.instancePath.split('/');
        paths.shift();
        const schema = error.schemaPath.split('/');
        schema.shift();
        if (schema.length > 0) {
            if (schema[0] === 'definitions') {
                schema.shift();
            } else if (model) {
                schema.unshift(model);
            }
        }

        console.log(chalk.underline(chalk.white(paths.length > 0 ? paths.join(' > ') : 'root')));
        console.log('  ', chalk.red('error'), ' ', error.message, error.keyword ? chalk.dim(` [${error.keyword}]`) : '');
        console.log('  ', chalk.gray(schema.join(' > ')));
        console.log('');
    }
}

export class CLIError extends Error {
    code;
    constructor(code: 'failed_to_parse_nango_yaml' | 'error_loading_nango_yaml', message?: string) {
        super(message || code);

        this.code = code;
    }
}
