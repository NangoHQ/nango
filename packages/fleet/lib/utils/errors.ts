import type { Jsonable } from '@nangohq/types';

type FleetErrorCode =
    | 'no_active_deployment'
    | 'deployment_creation_failed'
    | 'deployment_get_active_failed'
    | 'deployment_not_found'
    | 'node_invalid_state_transition'
    | 'node_not_found'
    | 'node_creation_failed'
    | 'node_search_failed'
    | 'node_transition_failed'
    | 'node_fail_failed'
    | 'node_register_failed'
    | 'node_delete_failed'
    | 'node_delete_non_terminated'
    | 'node_config_override_creation_failed'
    | 'node_config_override_update_failed'
    | 'node_config_override_reset_image_failed'
    | 'node_config_override_not_found'
    | 'node_config_override_delete_failed'
    | 'node_config_override_search_failed'
    | 'supervisor_search_nodes_failed'
    | 'supervisor_unknown_action'
    | 'supervisor_tick_failed'
    | 'fleet_misconfigured'
    | 'fleet_rollout_invalid_image'
    | 'fleet_rollout_image_not_found'
    | 'fleet_node_not_ready_timeout'
    | 'fleet_node_url_not_found'
    | 'fleet_node_outdate_failed'
    | 'fleet_cannot_acquire_lock'
    | 'fleet_lock_timeout'
    | 'fleet_lock_error'
    | 'local_runner_start_failed';

export class FleetError extends Error {
    public readonly context?: Jsonable;

    constructor(code: FleetErrorCode, options: { cause?: unknown; context?: Jsonable } = {}) {
        const { cause, context } = options;
        super(code, { cause });
        this.name = this.constructor.name;
        this.context = context;
    }
}
