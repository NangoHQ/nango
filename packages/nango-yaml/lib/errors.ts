export class ParserError {
    code;
    message: string;
    path: string;

    constructor({ code, message, path }: { code: string; message: string; path: string[] }) {
        this.code = code;
        this.message = message;
        this.path = path.join(' > ');
    }
}

export class ParserErrorInvalidModelName extends ParserError {
    constructor(options: { model: string; path: string[] }) {
        super({
            code: 'invalid_model_name',
            message: `Model "${options.model}" contains invalid characters`,
            path: options.path
        });
    }
}

export class ParserErrorDuplicateModel extends ParserError {
    constructor(options: { model: string; path: string[] }) {
        super({
            code: 'duplicate_model',
            message: `Model "${options.model}" is used multiple times. Please make sure all models are unique within an integration.`,
            path: options.path
        });
    }
}

export class ParserErrorModelNotFound extends ParserError {
    constructor(options: { model: string; path: string[] }) {
        super({
            code: 'model_not_found',
            message: `Model "${options.model}" does not exists`,
            path: options.path
        });
    }
}

export class ParserErrorMissingId extends ParserError {
    constructor(options: { model: string; path: string[] }) {
        super({
            code: 'model_missing_id',
            message: `Model "${options.model}" doesn't have an id field. This is required to be able to uniquely identify the data record`,
            path: options.path
        });
    }
}

export class ParserErrorDuplicateEndpoint extends ParserError {
    constructor(options: { endpoint: string; path: string[] }) {
        super({
            code: 'duplicate_endpoint',
            message: `Endpoint "${options.endpoint}" is used multiple times. Please make sure all endpoints are unique.`,
            path: options.path
        });
    }
}

export class ParserErrorEndpointsMismatch extends ParserError {
    constructor(options: { syncName: string; path: string[] }) {
        super({
            code: 'endpoints_mismatch',
            message: `The number of endpoints doesn't match the number of models returned by "${options.syncName}". The endpoints to model should match 1 to 1.`,
            path: options.path
        });
    }
}

export class ParserErrorInvalidRuns extends ParserError {
    constructor(options: { message: string; path: string[] }) {
        super({
            code: 'invalid_runs',
            message: options.message,
            path: options.path
        });
    }
}

export class ParserErrorModelIsLiteral extends ParserError {
    constructor(options: { model: string; path: string[] }) {
        super({
            code: 'model_is_literal',
            message: `A literal type "${options.model}" was parsed, which might be an error. It is recommended to use a Model for input and output.`,
            path: options.path
        });
    }
}

export class ParserErrorCycle extends ParserError {
    constructor(options: { stack: Set<string> }) {
        const arr = Array.from(options.stack);
        const start = arr.shift();
        const end = arr.pop();
        super({
            code: 'cyclic_model',
            message: `Cyclic import ${start}->${end}`,
            path: Array.from(options.stack)
        });
    }
}

export class ParserErrorExtendsNotFound extends ParserError {
    constructor(options: { model: string; inherit: string; path: string[] }) {
        super({
            code: 'model_extends_not_found',
            message: `Model "${options.model}" is extending "${options.inherit}", but it does not exists`,
            path: options.path
        });
    }
}

export class ParserErrorTypeSyntax extends ParserError {
    constructor(options: { value: string; path: string[] }) {
        super({
            code: 'type_syntax_error',
            message: `Type "${options.value}" contains some unsupported typescript syntax, please use allowed syntax. Documentation: https://docs.nango.dev/reference/integration-configuration#model-types`,
            path: options.path
        });
    }
}

export class ParserErrorBothPostConnectionScriptsAndOnEventsPresent extends ParserError {
    constructor(options: { path: string[] }) {
        super({
            code: 'both_post_connection_scripts_and_on_events_present',
            message: `Both post-connection-scripts and on-events are present. Only one of them can be used at a time.`,
            path: options.path
        });
    }
}
