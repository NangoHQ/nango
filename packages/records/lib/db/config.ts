import { envs } from '../env.js';
import type { Knex } from 'knex';

export const schema = envs.RECORDS_DATABASE_SCHEMA;
const databaseUrl =
    envs.RECORDS_DATABASE_URL ||
    envs.NANGO_DATABASE_URL ||
    `postgres://${encodeURIComponent(envs.NANGO_DB_USER)}:${encodeURIComponent(envs.NANGO_DB_PASSWORD)}@${envs.NANGO_DB_HOST}:${envs.NANGO_DB_PORT}/${envs.NANGO_DB_NAME}?application_name=${envs.NANGO_DB_APPLICATION_NAME}`;
const runningMigrationOnly = process.argv.some((v) => v === 'migrate:latest');
const isJS = !runningMigrationOnly;

const config: Knex.Config = {
    client: 'postgres',
    connection: {
        connectionString: databaseUrl,
        statement_timeout: 60000
    },
    searchPath: schema,
    pool: { min: 2, max: 50 },
    migrations: {
        extension: isJS ? 'js' : 'ts',
        directory: 'migrations',
        tableName: 'migrations',
        loadExtensions: [isJS ? '.js' : '.ts'],
        schemaName: schema
    }
};

// This config is optional because it can create issues if we have 2 pools connected to the same URL
const configRead: Knex.Config | undefined = envs.RECORDS_DATABASE_READ_URL
    ? {
          ...config,
          connection: {
              connectionString: envs.RECORDS_DATABASE_READ_URL,
              statement_timeout: 60000
          }
      }
    : undefined;

export { config, configRead };
