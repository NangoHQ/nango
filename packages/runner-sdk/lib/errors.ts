export abstract class SDKError extends Error {
    abstract code: string;
    payload: Record<string, unknown>;

    constructor(payload?: Record<string, unknown>) {
        super();
        this.payload = payload || {};
    }
}

export class AbortedSDKError extends SDKError {
    code = 'script_aborted';
}

export class UnknownProviderSDKError extends SDKError {
    code = 'unknown_provider_template_in_config';
}

export class InvalidRecordSDKError extends SDKError {
    code = 'invalid_sync_record';
}

export class InvalidActionInputSDKError extends SDKError {
    code = 'invalid_action_input';
}

export class InvalidActionOutputSDKError extends SDKError {
    code = 'invalid_action_output';
}

/**
 * For external use only
 */
export class ActionError<T = Record<string, unknown>> extends Error {
    type: string;
    payload?: Record<string, unknown>;

    constructor(payload?: T) {
        super();
        this.type = 'action_script_runtime_error';
        if (payload) {
            this.payload = payload;
        }
    }
}
