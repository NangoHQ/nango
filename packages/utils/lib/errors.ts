import { serializeError } from 'serialize-error';
import * as Sentry from '@sentry/node';
import { getLogger } from './logger.js';
import { NANGO_VERSION } from './version.js';

/**
 * Transform any Error or primitive to a json object
 */
export function errorToObject(err: unknown) {
    return serializeError(err, { maxDepth: 5 });
}

/**
 * Transform any Error or primitive to a string
 */
export function stringifyError(err: unknown, opts?: { pretty?: boolean; stack?: boolean }) {
    return JSON.stringify(serializeError(err), ['name', 'message', ...(opts?.stack ? ['stack', 'cause'] : [])], opts?.pretty ? 2 : undefined);
}

export function initSentry({ dsn, hash, applicationName }: { dsn: string | undefined; hash?: string | undefined; applicationName: string }) {
    Sentry.init({
        dsn: dsn || '',
        sampleRate: 1,
        skipOpenTelemetrySetup: true, // If false or not set, sentry is breaking our otel setup for logs export
        enabled: dsn ? true : false,
        release: `${NANGO_VERSION}@${hash || 'no_hash'}`,
        serverName: applicationName,
        maxBreadcrumbs: 10
    });
    if (dsn) {
        logger.info('Sentry configured');
    }
}

const logger = getLogger('[err]');
export function report(err: unknown, extra?: Record<string, string | number | null | undefined>) {
    logger.error(err as any, extra);

    Sentry.withScope((scope) => {
        if (extra) {
            scope.setExtras(extra);
        }

        Sentry.captureException(err);
    });
}
