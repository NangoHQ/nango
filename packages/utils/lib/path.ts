import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
export const projectRoot = path.join(__filename, '../../../..');
