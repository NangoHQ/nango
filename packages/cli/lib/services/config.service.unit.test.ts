import path from 'node:path';
import yaml from 'js-yaml';
import stripAnsi from 'strip-ansi';
import { afterEach, describe, expect, it, vi } from 'vitest';
import { parse, validateYaml } from './config.service';
import { CLIError } from '../utils/errors.js';

function cleanLog(log: any) {
    return typeof log === 'string' ? stripAnsi(log) : log;
}
describe('load', () => {
    // Not the best but until we have a logger it will work
    const consoleMock = vi.spyOn(console, 'log').mockImplementation(() => undefined);
    afterEach(() => {
        consoleMock.mockReset();
    });

    it('should parse a nango.yaml file that is version 1 as expected', () => {
        const res = parse(path.resolve(__dirname, `../../fixtures/nango-yaml/v1/valid`));
        if (res.isErr()) {
            throw res.error;
        }

        expect(res.value.parsed).toMatchSnapshot();
    });

    it('should parse a nango.yaml file that is version 2 as expected', () => {
        const res = parse(path.resolve(__dirname, `../../fixtures/nango-yaml/v2/valid`));
        if (res.isErr()) {
            throw res.error;
        }

        expect(res.value.parsed).toMatchSnapshot();
    });

    it('should throw a validation error on a nango.yaml file that is not formatted correctly -- missing endpoint', () => {
        const acc: string[] = [];
        consoleMock.mockImplementation((m) => acc.push(cleanLog(m)));

        const res = parse(path.resolve(__dirname, `../../fixtures/nango-yaml/v2/invalid.1`));
        if (res.isOk()) {
            expect(res).toBeInstanceOf(Error);
            return;
        }

        expect(res.error).toStrictEqual(new CLIError('failed_to_parse_nango_yaml', 'Your nango.yaml contains some errors'));
        expect(acc.join('')).toContain('An endpoint property is required to specify how to retrieve the data from the sync');
    });

    it('should throw a validation error on a nango.yaml file that is not formatted correctly -- webhook subscriptions are not allowed in an action', () => {
        const acc: string[] = [];
        consoleMock.mockImplementation((m) => acc.push(stripAnsi(m)));

        const res = parse(path.resolve(__dirname, `../../fixtures/nango-yaml/v2/invalid.2`));
        if (res.isOk()) {
            expect(res).toBeInstanceOf(Error);
            return;
        }

        expect(res.error).toStrictEqual(new CLIError('failed_to_parse_nango_yaml', 'Your nango.yaml contains some errors'));
        expect(acc.join('')).toContain('additionalProperty: webhook-subscription');
    });
});

describe('validate', () => {
    it('should validate empty yaml', () => {
        const res = validateYaml(yaml.load(``));
        expect(res).toStrictEqual([
            {
                msg: 'Invalid file format, you should have at least an `integrations` property at the root level. Check our documentation https://docs.nango.dev/reference/integration-configuration'
            }
        ]);
    });

    it('should require sync to be object', () => {
        const res = validateYaml(
            yaml.load(`
integrations:
  test:
    syncs:
      foobar:
`)
        );
        expect(res).toStrictEqual([
            {
                code: 'type',
                msg: 'must be object',
                params: { type: 'object' },
                path: '/integrations/test/syncs/foobar'
            }
        ]);
    });

    it('should require some fields in sync', () => {
        const res = validateYaml(
            yaml.load(`
integrations:
  test:
    syncs:
      foobar:
        input: boolean
        runs: every day
`)
        );
        expect(res).toMatchObject([
            {
                msg: 'An endpoint property is required to specify how to retrieve the data from the sync.',
                path: '/integrations/test/syncs/foobar'
            },
            {
                msg: 'An output property is required to specify what is the outcome of the sync.',
                path: '/integrations/test/syncs/foobar'
            }
        ]);
    });

    it('should disallow extra properties', () => {
        const res = validateYaml(
            yaml.load(`
integrations:
  test:
    syncs:
      foobar:
        runs: every day
        endpoint: GET /test
        output: boolean
        iinput: boolean
`)
        );
        expect(res).toStrictEqual([
            {
                code: 'additionalProperties',
                msg: 'must NOT have additional properties',
                params: { additionalProperty: 'iinput' },
                path: '/integrations/test/syncs/foobar'
            }
        ]);
    });

    it('should enforce endpoint format', () => {
        const res = validateYaml(
            yaml.load(`
integrations:
  test:
    syncs:
      foobar:
        runs: every day
        endpoint: /te df /+-st
        output: boolean
`)
        );
        expect(res).toMatchObject([
            {
                msg: `endpoint must be a URL (or an array of URLs) with an HTTP verb, i.e:
- endpoint: GET /tickets
- endpoint:
    method: GET
    path: /tickets`,
                path: '/integrations/test/syncs/foobar/endpoint'
            }
        ]);
    });

    it('should prevent incorrect HTTP verb for sync endpoint ', () => {
        const res = validateYaml(
            yaml.load(`
integrations:
  test:
    sync:
      foobar:
        endpoint: DELETE /test
        output: boolean
`)
        );
        expect(res).toMatchObject([]);
    });

    it('should allow endpoint format for action', () => {
        const res = validateYaml(
            yaml.load(`
integrations:
  test:
    actions:
      foobar:
        endpoint: DELETE /test
        output: boolean
`)
        );
        expect(res).toMatchObject([]);
    });
});
