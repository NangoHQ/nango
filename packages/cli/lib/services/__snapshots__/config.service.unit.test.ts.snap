// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`load > should parse a nango.yaml file that is version 1 as expected 1`] = `
{
  "integrations": [
    {
      "actions": [
        {
          "description": "",
          "endpoint": null,
          "input": null,
          "name": "github-write-action",
          "output": [
            "Anonymous_demogithubintegration_sync_githubwriteaction_output",
          ],
          "scopes": [],
          "type": "action",
          "usedModels": [
            "Anonymous_demogithubintegration_sync_githubwriteaction_output",
          ],
          "version": "",
        },
      ],
      "onEventScripts": {
        "post-connection-creation": [],
        "pre-connection-deletion": [],
      },
      "providerConfigKey": "demo-github-integration",
      "syncs": [
        {
          "auto_start": true,
          "description": "",
          "endpoints": [],
          "input": null,
          "name": "github-issue-example",
          "output": [
            "GithubIssue",
          ],
          "runs": "every half hour",
          "scopes": [],
          "sync_type": "incremental",
          "track_deletes": false,
          "type": "sync",
          "usedModels": [
            "GithubIssue",
          ],
          "version": "",
          "webhookSubscriptions": [],
        },
      ],
    },
  ],
  "models": Map {
    "GithubIssue" => {
      "fields": [
        {
          "array": false,
          "name": "id",
          "optional": false,
          "tsType": true,
          "value": "number",
        },
        {
          "array": false,
          "name": "owner",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "repo",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "issue_number",
          "optional": false,
          "tsType": true,
          "value": "number",
        },
        {
          "array": false,
          "name": "title",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "author",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "author_id",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "state",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "date_created",
          "optional": false,
          "tsType": true,
          "value": "Date",
        },
        {
          "array": false,
          "name": "date_last_modified",
          "optional": false,
          "tsType": true,
          "value": "Date",
        },
        {
          "array": false,
          "name": "body",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
      ],
      "name": "GithubIssue",
    },
    "Anonymous_demogithubintegration_sync_githubwriteaction_output" => {
      "fields": [
        {
          "array": false,
          "name": "output",
          "optional": false,
          "tsType": true,
          "value": "boolean",
        },
      ],
      "isAnon": true,
      "name": "Anonymous_demogithubintegration_sync_githubwriteaction_output",
    },
  },
  "yamlVersion": "v1",
}
`;

exports[`load > should parse a nango.yaml file that is version 2 as expected 1`] = `
{
  "integrations": [
    {
      "actions": [
        {
          "description": "Get a GitHub issue.",
          "endpoint": {
            "method": "GET",
            "path": "/ticketing/tickets/{GithubCreateIssueInput:id}",
          },
          "input": null,
          "name": "github-get-issue",
          "output": [
            "GithubIssueAction",
          ],
          "scopes": [
            "repo:read",
          ],
          "type": "action",
          "usedModels": [
            "GithubIssueAction",
          ],
          "version": "",
        },
        {
          "description": "Creates a GitHub issue.",
          "endpoint": {
            "method": "POST",
            "path": "/ticketing/tickets",
          },
          "input": "GithubCreateIssueInput",
          "name": "github-create-issue",
          "output": [
            "GithubCreateOutput",
          ],
          "scopes": [
            "repo:write",
          ],
          "type": "action",
          "usedModels": [
            "GithubCreateOutput",
            "GithubIssue",
            "GithubCreateIssueInput",
          ],
          "version": "",
        },
        {
          "description": "Deletes a GitHub issue.",
          "endpoint": {
            "method": "DELETE",
            "path": "/ticketing/tickets/{GithubIssue:id}",
          },
          "input": null,
          "name": "github-delete-issue",
          "output": [
            "Anonymous_demogithubintegration_action_githubdeleteissue_output",
          ],
          "scopes": [
            "repo:write",
          ],
          "type": "action",
          "usedModels": [
            "Anonymous_demogithubintegration_action_githubdeleteissue_output",
          ],
          "version": "",
        },
      ],
      "onEventScripts": {
        "post-connection-creation": [],
        "pre-connection-deletion": [],
      },
      "providerConfigKey": "demo-github-integration",
      "syncs": [
        {
          "auto_start": true,
          "description": "Sync github issues continuously from public repos",
          "endpoints": [
            {
              "method": "GET",
              "path": "/ticketing/tickets",
            },
          ],
          "input": null,
          "name": "github-issue-example",
          "output": [
            "GithubIssue",
          ],
          "runs": "every half hour",
          "scopes": [
            "public_repo",
          ],
          "sync_type": "incremental",
          "track_deletes": false,
          "type": "sync",
          "usedModels": [
            "GithubIssue",
          ],
          "version": "",
          "webhookSubscriptions": [
            "issue.creation",
          ],
        },
        {
          "auto_start": true,
          "description": "Sync github issues continuously from public repos example two",
          "endpoints": [
            {
              "method": "GET",
              "path": "/ticketing/tickets-two",
            },
          ],
          "input": null,
          "name": "github-issue-example-two",
          "output": [
            "GithubIssue2",
          ],
          "runs": "every hour",
          "scopes": [
            "public_repo",
          ],
          "sync_type": "incremental",
          "track_deletes": false,
          "type": "sync",
          "usedModels": [
            "GithubIssue2",
          ],
          "version": "",
          "webhookSubscriptions": [],
        },
        {
          "auto_start": true,
          "description": "Sync github issues to multiple models",
          "endpoints": [
            {
              "method": "GET",
              "path": "/ticketing/ticket",
            },
            {
              "method": "GET",
              "path": "/ticketing/pr",
            },
          ],
          "input": null,
          "name": "github-multiple-models",
          "output": [
            "GithubIssue3",
            "GithubPR",
          ],
          "runs": "every 5 minutes",
          "scopes": [],
          "sync_type": "full",
          "track_deletes": false,
          "type": "sync",
          "usedModels": [
            "GithubIssue3",
            "GithubPR",
          ],
          "version": "",
          "webhookSubscriptions": [],
        },
      ],
    },
  ],
  "models": Map {
    "GithubIssue" => {
      "fields": [
        {
          "array": false,
          "name": "id",
          "optional": false,
          "tsType": true,
          "value": "number",
        },
        {
          "array": false,
          "name": "owner",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "repo",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "issue_number",
          "optional": false,
          "tsType": true,
          "value": "number",
        },
        {
          "array": false,
          "name": "title",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "author",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "author_id",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "state",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "date_created",
          "optional": false,
          "tsType": true,
          "value": "Date",
        },
        {
          "array": false,
          "name": "date_last_modified",
          "optional": false,
          "tsType": true,
          "value": "Date",
        },
        {
          "array": false,
          "name": "body",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
      ],
      "name": "GithubIssue",
    },
    "GithubIssueAction" => {
      "fields": [
        {
          "array": false,
          "name": "id",
          "optional": false,
          "tsType": true,
          "value": "number",
        },
        {
          "array": false,
          "name": "owner",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "repo",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "issue_number",
          "optional": false,
          "tsType": true,
          "value": "number",
        },
        {
          "array": false,
          "name": "title",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "author",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "author_id",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "state",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "date_created",
          "optional": false,
          "tsType": true,
          "value": "Date",
        },
        {
          "array": false,
          "name": "date_last_modified",
          "optional": false,
          "tsType": true,
          "value": "Date",
        },
        {
          "array": false,
          "name": "body",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
      ],
      "name": "GithubIssueAction",
    },
    "GithubIssue2" => {
      "fields": [
        {
          "array": false,
          "name": "id",
          "optional": false,
          "tsType": true,
          "value": "number",
        },
        {
          "array": false,
          "name": "owner",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "repo",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "issue_number",
          "optional": false,
          "tsType": true,
          "value": "number",
        },
        {
          "array": false,
          "name": "title",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "author",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "author_id",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "state",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "date_created",
          "optional": false,
          "tsType": true,
          "value": "Date",
        },
        {
          "array": false,
          "name": "date_last_modified",
          "optional": false,
          "tsType": true,
          "value": "Date",
        },
        {
          "array": false,
          "name": "body",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
      ],
      "name": "GithubIssue2",
    },
    "GithubIssue3" => {
      "fields": [
        {
          "array": false,
          "name": "id",
          "optional": false,
          "tsType": true,
          "value": "number",
        },
        {
          "array": false,
          "name": "owner",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "repo",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "issue_number",
          "optional": false,
          "tsType": true,
          "value": "number",
        },
        {
          "array": false,
          "name": "title",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "author",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "author_id",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "state",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "date_created",
          "optional": false,
          "tsType": true,
          "value": "Date",
        },
        {
          "array": false,
          "name": "date_last_modified",
          "optional": false,
          "tsType": true,
          "value": "Date",
        },
        {
          "array": false,
          "name": "body",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
      ],
      "name": "GithubIssue3",
    },
    "GithubCreateIssueInput" => {
      "fields": [
        {
          "array": false,
          "name": "id",
          "optional": false,
          "tsType": true,
          "value": "number",
        },
        {
          "array": false,
          "name": "owner",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "repo",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "issue_number",
          "optional": false,
          "tsType": true,
          "value": "number",
        },
        {
          "array": false,
          "name": "title",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "author",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "author_id",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "state",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "date_created",
          "optional": false,
          "tsType": true,
          "value": "Date",
        },
        {
          "array": false,
          "name": "date_last_modified",
          "optional": false,
          "tsType": true,
          "value": "Date",
        },
        {
          "array": false,
          "name": "body",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
      ],
      "name": "GithubCreateIssueInput",
    },
    "GithubCreateOutput" => {
      "fields": [
        {
          "array": false,
          "model": true,
          "name": "result",
          "optional": false,
          "value": "GithubIssue",
        },
      ],
      "name": "GithubCreateOutput",
    },
    "GithubPR" => {
      "fields": [
        {
          "array": false,
          "name": "id",
          "optional": false,
          "tsType": true,
          "value": "number",
        },
        {
          "array": false,
          "name": "owner",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "repo",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "issue_number",
          "optional": false,
          "tsType": true,
          "value": "number",
        },
        {
          "array": false,
          "name": "title",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "author",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "author_id",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "state",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
        {
          "array": false,
          "name": "date_created",
          "optional": false,
          "tsType": true,
          "value": "Date",
        },
        {
          "array": false,
          "name": "date_last_modified",
          "optional": false,
          "tsType": true,
          "value": "Date",
        },
        {
          "array": false,
          "name": "body",
          "optional": false,
          "tsType": true,
          "value": "string",
        },
      ],
      "name": "GithubPR",
    },
    "Anonymous_demogithubintegration_action_githubdeleteissue_output" => {
      "fields": [
        {
          "array": false,
          "name": "output",
          "optional": false,
          "tsType": true,
          "value": "boolean",
        },
      ],
      "isAnon": true,
      "name": "Anonymous_demogithubintegration_action_githubdeleteissue_output",
    },
  },
  "yamlVersion": "v2",
}
`;
